File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> left(nums.size());
        int pd = 1;
        for(int i=0; i<nums.size(); i++){
            left[i] = pd;
            pd *= nums[i];
        }

        vector<int> right(nums.size());
        pd=1;
        for(int i = nums.size()-1; i >= 0; i--){
            right[i] = pd;
            pd *= nums[i];
        }

        vector<int> ans;
        for(int i=0; i<nums.size(); i++){
            ans.push_back(left[i]*right[i]);
        }

        return ans;
    }
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int prod = 1;
        int zeroCount = 0;
        
        for (int x : nums) {
            if (x != 0) prod *= x;
            else zeroCount++;
        }

        int[] ans = new int[nums.length];
        for (int i = 0; i < nums.length; i++) {
            if (zeroCount > 1) ans[i] = 0;
            else if (zeroCount == 1) ans[i] = (nums[i] == 0) ? prod : 0;
            else ans[i] = prod / nums[i];
        }
        return ans;
    }
}
